src/string-view.h

```diff
@@ -0,0 +1,259 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef WABT_STRING_VIEW_H_
+#define WABT_STRING_VIEW_H_
+
+#include <cassert>
+#include <functional>
+#include <iterator>
+#include <string>
+
+#include "hash-util.h"
+
+namespace wabt {
+
+// This is a simplified implemention of C++17's basic_string_view template.
+//
+// Missing features:
+// * Not a template
+// * No allocator support
+// * Some functions are not implemented
+// * Asserts instead of exceptions
+
+class string_view {
+ public:
+  typedef std::char_traits<char> traits_type;
+  typedef char value_type;
+  typedef char* pointer;
+  typedef const char* const_pointer;
+  typedef char& reference;
+  typedef const char& const_reference;
+  typedef const char* const_iterator;
+  typedef const_iterator iterator;
+  typedef std::reverse_iterator<const_iterator> const_reverse_iterator;
+  typedef const_reverse_iterator reverse_iterator;
+  typedef std::size_t size_type;
+  typedef std::ptrdiff_t difference_type;
+  static const size_type npos = size_type(-1);
+
+  // construction and assignment
+  string_view() noexcept;
+  string_view(const string_view&) noexcept = default;
+  string_view& operator=(const string_view&) noexcept = default;
+  string_view(const std::string& str) noexcept;
+  string_view(const char* str);
+  string_view(const char* str, size_type len);
+
+  // iterator support
+  const_iterator begin() const noexcept;
+  const_iterator end() const noexcept;
+  const_iterator cbegin() const noexcept;
+  const_iterator cend() const noexcept;
+  const_reverse_iterator rbegin() const noexcept;
+  const_reverse_iterator rend() const noexcept;
+  const_reverse_iterator crbegin() const noexcept;
+  const_reverse_iterator crend() const noexcept;
+
+  // capacity
+  size_type size() const noexcept;
+  size_type length() const noexcept;
+  size_type max_size() const noexcept;
+  bool empty() const noexcept;
+
+  // element access
+  const_reference operator[](size_type pos) const;
+  const_reference at(size_type pos) const;
+  const_reference front() const;
+  const_reference back() const;
+  const_pointer data() const noexcept;
+
+  // modifiers
+  void remove_prefix(size_type n);
+  void remove_suffix(size_type n);
+  void swap(string_view& s) noexcept;
+
+  // string operations
+  explicit operator std::string() const;
+  std::string to_string() const;
+  size_type copy(char* s, size_type n, size_type pos = 0) const;
+  string_view substr(size_type pos = 0, size_type n = npos) const;
+  int compare(string_view s) const noexcept;
```

Comment:
This matches the definition here: http://en.cppreference.com/w/cpp/header/string_view